src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs

```diff
@@ -80,9 +190,41 @@ private static void ThrowLockTakenException()
         =========================================================================*/
         public static bool TryEnter(object obj)
         {
-            bool lockTaken = false;
-            TryEnter(obj, 0, ref lockTaken);
-            return lockTaken;
+            ArgumentNullException.ThrowIfNull(obj, null);
+
+            EnterHelperResult tryEnterResult = TryEnter_FastPath_WithTimeout(obj, 0);
+            if (tryEnterResult == EnterHelperResult.Entered)
+            {
+                return true;
+            }
+            else if (tryEnterResult == EnterHelperResult.Contention)
+            {
+                return false;
+            }
+
+            return TryEnter_Slowpath(obj);
+        }
+
+        private static void TryEnter_Timeout_WithLockTaken(object obj, int millisecondsTimeout, ref bool lockTaken)
+        {
+            if (millisecondsTimeout >= -1)
+            {
+                EnterHelperResult tryEnterResult = TryEnter_FastPath_WithTimeout(obj, millisecondsTimeout);
+                if (tryEnterResult == EnterHelperResult.Entered)
+                {
+                    lockTaken = true;
+                    return;
+                }
+                else if (millisecondsTimeout == 0 && (tryEnterResult == EnterHelperResult.Contention))
+                {
+                    return;
+                }
+            }
+
+            if (TryEnter_Slowpath(obj, millisecondsTimeout))
```

Comment:
Wouldn't a throw-helper be more efficient here? It would allow this block to be marked as cold-code.